src/plugins/auto/executable_network.cpp

```diff
@@ -137,10 +137,12 @@ void MultiDeviceExecutableNetwork::GenerateWorkers(const std::string& device, co
                     // let's try to pop a task, as we know there is at least one idle request, schedule if succeeded
                     // if no device-agnostic tasks, let's try pop the device specific task, schedule if succeeded
                     Task t;
-                    if (_inferPipelineTasks.try_pop(t))
-                        ScheduleToWorkerInferRequest(std::move(t));
-                    else if (_inferPipelineTasksDeviceSpecific[device]->try_pop(t))
-                        ScheduleToWorkerInferRequest(std::move(t), device);
+                    do {
+                        _inferPipelineTasks.try_pop(t);
+                    } while (t && ScheduleToWorkerInferRequest(std::move(t)));
+                    do {
+                        _inferPipelineTasksDeviceSpecific[device]->try_pop(t);
```

Comment:
from what I see the solution is to continue popping the requests from the queue, while **being in the callback** of the first request. May this result in heavy delaying of the completion notification?
Also with @apankratovantonp

Reply:
Seems that this code can be reduced to spinlock on idle worker queue. What if the destructor will ~MultiDeviceExecutableNetwork will be called before src/plugins/auto/executable_network.cpp:140. So the idle request queue will always be empty and that will lead to deadlock in ~MultiDeviceExecutableNetwork destructor.

Reply:
> from what I see the solution is to continue popping the requests from the queue, while **being in the callback** of the first request. May this result in heavy delaying of the completion notification? Also with @apankratovantonp

I think generally it will not bring heavy delay of completion notification, only during the transition stage maybe introduce some more delay on the remaining cpu helper requests, but comparing with the batch timeout, this is minor, as batch async request will return immediately.

Reply:
@songbell There is no spin_lock that I have mentioned before. And destructor should not halt.